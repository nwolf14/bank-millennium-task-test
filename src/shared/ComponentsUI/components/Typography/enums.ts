export enum Variant {
  h1 = 'h1',
  h2 = 'h2',
  h3 = 'h3',
  h4 = 'h4',
  h5 = 'h5',
  h6 = 'h6',
  text = 'p',
  smallText = 'small',
}

export enum FontStyle {
  italic = 'italic',
  normal = 'normal',
}

export enum FontWeight {
  black = 'black',
  bold = 'bold',
  extraBold = 'extra-bold',
  light = 'light',
  medium = 'medium',
  regular = 'regular',
  semiBold = 'semi-bold',
}
