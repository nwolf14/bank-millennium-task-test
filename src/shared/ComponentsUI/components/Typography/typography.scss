.Typography {
  @mixin typography-styles($font-size, $line-height, $margin-bottom) {
    display: block;
    font-size: $font-size;
    font-weight: 400;
    line-height: $line-height;
    margin: 0;
    margin-bottom: $margin-bottom;
    margin-inline-end: 0;
    margin-inline-start: 0;
    padding: 0;
  }

  @mixin heading() {
    font-family: 'Nunito', sans-serif;
  }

  @mixin standard() {
    font-family: 'Rubik', sans-serif;
  }

  &--h1 {
    @include heading;
    @include typography-styles(96px, 96px, 35px);
  }

  &--h2 {
    @include heading;
    @include typography-styles(60px, 60px, 20px);
  }

  &--h3 {
    @include heading;
    @include typography-styles(48px, 60px, 16px);
  }

  &--h4 {
    @include heading;
    @include typography-styles(34px, 50px, 12px);
  }

  &--h5 {
    @include heading;
    @include typography-styles(24px, 28px, 8px);
  }

  &--h6 {
    @include heading;
    @include typography-styles(20px, 28px, 6px);
  }

  &--text {
    @include standard;
    @include typography-styles(16px, 20px, 4px);
  }

  &--small-text {
    @include standard;
    @include typography-styles(12px, 16px, 2px);
  }

  &--normal {
    font-style: normal;
  }

  &--italic {
    font-style: italic;
  }

  &--light {
    font-weight: 300;
  }

  &--regular {
    font-weight: 400;
  }

  &--medium {
    font-weight: 500;
  }

  &--semi-bold {
    font-weight: 600;
  }

  &--bold {
    font-weight: 700;
  }

  &--extra-bold {
    font-weight: 800;
  }

  &--black {
    font-weight: 900;
  }
}
