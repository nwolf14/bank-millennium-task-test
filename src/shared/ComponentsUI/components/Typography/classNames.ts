import mapValues from 'lodash/mapValues';
import { FontStyle, FontWeight } from './enums';

export const className = 'Typography';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      h1: `${className}--h1`,
      h2: `${className}--h2`,
      h3: `${className}--h3`,
      h4: `${className}--h4`,
      h5: `${className}--h5`,
      h6: `${className}--h6`,
      p: `${className}--text`,
      small: `${className}--small-text`,
      ...mapValues(FontStyle, (fontStyle) => `${className}--${fontStyle}`),
      ...mapValues(FontWeight, (fontWeight) => `${className}--${fontWeight}`),
    },
  },
};
