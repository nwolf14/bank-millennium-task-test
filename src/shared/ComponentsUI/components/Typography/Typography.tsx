import classnames from 'classnames';
import { camelCase } from 'lodash';
import { createElement, FC, HTMLAttributes, ReactNode } from 'react';

// components
import { E2EAttribute } from '../../../E2EDataAttributes/enums';

// hooks
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameTypography, classNames } from './classNames';
import { DATA_TEST_PREFIX } from '../../../E2EDataAttributes/constants';
import { FontStyle, FontWeight, Variant } from './enums';

// styles
import './typography.scss';

export type TProps = {
  attributeValue?: string;
  children?: ReactNode;
  fontStyle?: FontStyle;
  fontWeight?: FontWeight;
  variant?: Variant;
} & HTMLAttributes<HTMLElement>;

const Typography: FC<TProps> = ({
  attributeValue = null,
  children,
  className = '',
  fontStyle = FontStyle.normal,
  fontWeight = FontWeight.regular,
  variant = Variant.text,
  ...restProps
}) => {
  const { classNamesWithTheme } = useTheme(classNames);

  if (!children) {
    return null;
  }

  return createElement(
    variant,
    {
      className: classnames(
        classNamesWithTheme[classNameTypography].name,
        className,
        classNamesWithTheme[classNameTypography].modificators[
          camelCase(fontWeight)
        ],
        classNamesWithTheme[classNameTypography].modificators[fontStyle],
        classNamesWithTheme[classNameTypography].modificators[
          camelCase(variant)
        ]
      ),
      [`${DATA_TEST_PREFIX}-${E2EAttribute.text}`]: attributeValue,
      ...restProps,
    },
    children
  );
};

export default Typography;
