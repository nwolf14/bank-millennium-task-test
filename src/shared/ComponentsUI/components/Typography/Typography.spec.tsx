import { render } from '@testing-library/react';

// components
import Typography from './Typography';

describe('Typography', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Typography />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Typography', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Typography>Text</Typography>);

    expect(asFragment()).toMatchSnapshot();
  });
});
