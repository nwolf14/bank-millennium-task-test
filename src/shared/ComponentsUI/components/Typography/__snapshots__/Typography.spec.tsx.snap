// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography matches snapshot 1`] = `<DocumentFragment />`;

exports[`Typography matches snapshot 2`] = `
<DocumentFragment>
  <p
    class="Typography Typography--regular Typography--normal Typography--text"
  >
    Text
  </p>
</DocumentFragment>
`;
